### PR TITLE
Separate snapshots logic, allow to wait for snapshots to be ready

### LINKED ISSUE #2286: GCP: Backup marked completed, even though snapshots are not ready yet
**What steps did you take and what happened:**
In our automation, I am creating backups on one cluster and then restore them on another within the same GCP project.
To backup, I use: 
```
velero backup create mybackup --storage-location gcp-s3 --include-namespaces mynamespace,myothernamespace --include-resources persistentvolumes --include-cluster-resources --snapshot-volumes --ttl 24h --wait
```
After some time the backup completes. However, when I check GCP -> Compute -> Snapshots, I see that not all snapshots are ready yet.
When restoring this 'completed' backup, I get some error messages,
so in my automation I included a step where I check all snapshots from the backup and only continue when all snapshots are marked as ready.

**What did you expect to happen:**
When I pass `--wait` to the backup command I expect the backup to be usable and all involved steps to be done before reporting a 'completed' status.

**The output of the following commands will help us better understand what's going on**:
`velero restore describe myrestore --details`
```
Name:         21-02-2020-mls-classic-platform-pv-backups-20200221084443
Namespace:    velero
Labels:       <none>
Annotations:  <none>

Phase:  PartiallyFailed (run 'velero restore logs 21-02-2020-mls-classic-platform-pv-backups-20200221084443' for more information)

Errors:
  Velero:     <none>
  Cluster:  error executing PVAction for persistentvolumes/pvc-442c6854-971d-4c0b-acff-ca4e840ccf0e: rpc error: code = Unknown desc = googleapi: Error 400: The resource 'projects/mls-classic-platform/global/snapshots/gke-mls-classic-platfo-pvc-ab9b674c-601f-411e-b565-4b244f1c9ce1' is not ready, resourceNotReady
  Namespaces: <none>

Backup:  21-02-2020-mls-classic-platform-pv-backups

Namespaces:
  Included:  *
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        nodes, events, events.events.k8s.io, backups.velero.io, restores.velero.io, resticrepositories.velero.io
  Cluster-scoped:  auto

Namespace mappings:  <none>

Label selector:  <none>

Restore PVs:  auto
```

**Environment:**

- Velero version (use `velero version`): 
```
Client:
	Version: v1.2.0
	Git commit: 5d008491bbf681658d3e372da1a9d3a21ca4c03c
Server:
	Version: v1.2.0 
```
- Velero features (use `velero client config get features`): None
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.2", GitCommit:"59603c6e503c87169aea6106f57b9f242f64df89", GitTreeState:"clean", BuildDate:"2020-01-18T23:30:10Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15+", GitVersion:"v1.15.7-gke.23", GitCommit:"06e05fd0390a51ea009245a90363f9161b6f2389", GitTreeState:"clean", BuildDate:"2020-01-17T23:10:45Z", GoVersion:"go1.12.12b4", Compiler:"gc", Platform:"linux/amd64"}
```
- Kubernetes installer & version: GKE
- Cloud provider or hardware configuration: GKE
- OS (e.g. from `/etc/os-release`): Ubuntu 18.04

### PR DESCRIPTION
**Pull request by boxcee**
Closes #2286

### CODE REVIEW

File: pkg/apis/velero/v1/snapshot.go [hunk 1]
```diff
@@ -31,36 +25,45 @@ type SnapshotSpec struct {
...
 	// VolumeIOPS is the optional value of provisioned IOPS for the
 	// disk/volume in the cloud provider API.
+	// +optional
 	VolumeIOPS *int64 `json:"volumeIOPS,omitempty"`
 }
 
 type SnapshotStatus struct {
 	// ProviderSnapshotID is the ID of the snapshot taken in the cloud
 	// provider API of this volume.
+	// +optional
```
Comment (nrb) [inline pkg/apis/velero/v1/snapshot.go]: I might be missing something, but why change this field to optional? Without it, we wouldn't be able to locate a volume snapshot.

File: pkg/apis/velero/v1/snapshot.go [hunk 2]
```diff
@@ -31,36 +25,45 @@ type SnapshotSpec struct {
...
 	BackupUID string `json:"backupUID"`
 
 	// Location is the name of the VolumeSnapshotLocation where this snapshot is stored.
+	// +optional
 	Location string `json:"location"`
 
 	// PersistentVolumeName is the Kubernetes name for the volume.
-	PersistentVolumeName string `json:persistentVolumeName`
+	// +optional
+	PersistentVolumeName string `json:"persistentVolumeName"`
```
Comment (nrb) [inline pkg/apis/velero/v1/snapshot.go]: Would at least one of `PersistentVolumeName` or `ProviderVolumeID` have to be populated?

File: pkg/apis/velero/v1/snapshot.go [hunk 3]
```diff
@@ -77,4 +80,37 @@ const (
...
+
+// Snapshot stores information about a persistent volume snapshot taken as
+// part of a Velero backup.
+type Snapshot struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
```
Comment (nrb) [inline pkg/apis/velero/v1/snapshot.go]: I don't think the entire spec can be optional - some minimum would need to be defined in order to know what needs to be snapshotted or restored.

**Comment by arianitu** [comment]
Does this PR still have plans to be merged or there's other plans for waiting on volume snapshots?

**Comment by skriss** [comment]
> Does this PR still have plans to be merged or there's other plans for waiting on volume snapshots?

I believe this PR extracted the Snapshot struct to be an API type, but it didn't actually implement the "wait for completion" logic, which will likely require plugin changes. I think it would make the most sense for someone to put up a design doc for this first, so we can get in sync on the approach before coding.

This is not currently prioritized in our roadmap, but is definitely something that makes sense for the project -- we're open to input & contributions.

**Comment by nrb** [comment]
Given the age of this, I'm going to close it out. I agree with the latest comment that I think it could use a design document to think through the implications.

I'm not at all opposed to it, because I think it'll provide a huge benefit to Velero and its community, but I think a design document is warranted for a change like this.

**Pull request closed without merging**